src/main/java/io/cdap/plugin/google/shopping/content/source/ShoppingContentClient.java

```diff
@@ -0,0 +1,159 @@
+      // Uses default credential is service account file is not provided.
+      initializer = GoogleCredential.getApplicationDefault();
+    } else {
+      initializer = loadCredential(serviceAccountPath, httpTransport);
+    }
+
+    ShoppingContent.Builder builder =
+        new ShoppingContent.Builder(httpTransport, JacksonFactory.getDefaultInstance(),
+            initializer);
+    this.shoppingContent = builder.build();
+  }
+
+  /**
+   * If true, we can call getNextPage().
+   * */
+  public boolean hasNextPage() throws IOException {
+    ShoppingContent.Products.List productsList = shoppingContent.products().list(merchantId)
+        .setIncludeInvalidInsertedItems(true);
+
+    if (end) {
+      return false;
+    }
+
+    currentPage = new ArrayList<>();
+
+    if (MAX_PRODUCTS > 0L) {
+      productsList.setMaxResults(MAX_PRODUCTS);
+    }
+
+    if (nextPageToken != null) {
+      productsList.setPageToken(nextPageToken);
+    }
+
+    ProductsListResponse page = productsList.execute();
+
+    if (page.getResources() == null || page.getResources().isEmpty()) {
+      LOGGER.info("No products found.");
+      return false;
+    }
+
+    for (Product product : page.getResources()) {
+      currentPage.add(product);
+    }
+
+    if (page.getNextPageToken() != null) {
+      nextPageToken = page.getNextPageToken();
+    } else {
+      nextPageToken = null;
+      end = true;
+    }
+
+    return true;
+  }
+
+  /**
+   * Lists a page of products given a merchant has uploaded to Google Shopping.
+   * Valid iff hasNextPage() returns true.
+   * @return A list of products given the merchant has uploaded to Google Shopping.
+   */
+  public List<Product> getNextPage() {
+    return currentPage;
+  }
+
+  /**
+   * Given the path to service account file, it loads a file named service-account.json and uses that json to
+   * authenticate with Google Shopping.
+   *
+   * @param serviceAccountPath The full path to the service account file.
+   * @param transport A HttpTransport.
+   * @return An authorized {@link Credential} object.
+   */
+  private Credential loadCredential(String serviceAccountPath, HttpTransport transport)
+      throws IOException {
+    File serviceAccountFile = new File(serviceAccountPath);
+
+    if (serviceAccountFile.exists()) {
+      LOGGER.info("Loading service account credentials.");
+
+      try (InputStream inputStream = new FileInputStream(serviceAccountFile)) {
+        GoogleCredential credential =
+            GoogleCredential
+                .fromStream(inputStream, transport, JacksonFactory.getDefaultInstance())
+                .createScoped(ShoppingContentScopes.all());
+
+        if (!credential.refreshToken()) {
+          LOGGER.debug("The service account access token could not be refreshed.");
+          LOGGER.debug("The service account key may have been deleted in the API Console.");
+          throw new IOException("Failed to refresh service account token");
```

Comment:
the log messages above are a better exception message because they tell the user what to do next. Let's remove the debug messages and change this to:

The service account access token could not be refreshed. Please ensure that the service account exists.